html-aam/scripts/html-aam-accname-utils.js

```diff
@@ -0,0 +1,338 @@
+const HtmlAamAccnameUtils = {
+  /*
+  Builds a collection of WPT subtests for verifying the computation steps
+  documented in https://www.w3.org/TR/html-aam-1.0/#accname-computation.
+
+  For each element in elements[], a collection of subtests will be generated to
+  test every potential name source in nameSources[].
+
+  Each entry in elements[] must have a key that is a valid CSS selector
+  conforming to either:
+
+  - `tagname`
+  - `tagname[attrname="attrvalue"]`
+
+  This key will be parsed to create the DOM element which will be used to
+  test the computed name.
+
+  Each entry in elements[] may have an optional properties object specifying
+  any of the following:
+
+  - children[]: elements to be appended as children of the test element. This
+    array may include the string 'textNode' to append a text node.
+  - parents[]: elements to be prepended as parents of the test element.
+  - attrs[]: attributes to be added to the element for every subtest. For
+    example, an `a` element may have an href attribute.
+  - specialContext: a string indicating a special case that requires complex
+    test markup. Currently supported values are 'area', 'ruby', 'rp', and 'rt'.
+
+  nameSources[] is necessarily order-sensitive, matching the computation steps
+  in HTML-AAM. Each string in nameSources[] is a unique identifier and should
+  have a matching key in the nameValues{} dictionary later in this file.
+
+  The first subtest generated will include *every* potential name source and
+  verify that the first source “wins”. Each subsequent subtest eliminates the
+  name source from the previous test, then performs the same verification.
+  In the final subtest, all potential name sources have been eliminated and the
+  computed name is expected to be empty.
+
+  The following example test configuration will generate 24 WPT subtests:
+  3 elements * 8 potential name sources = 24 subtests
+
+  Example:
+    const testConfig = {
+      elements: {
+        'input[type="text"]': { attrs: ['value'] },
+        'input[type="search"]': { attrs: ['value'] },
+        'textarea': { children: ['textNode'] }
+      },
+      nameSources: [
+        'from 2 aria-labelledby refs',
+        'from 1 aria-labelledby ref',
+        'from aria-label',
+        'from 2 labels',
+        'from 1 label',
+        'from title',
+        'from placeholder',
+        'to be empty'
+      ]
+    }
+    HtmlAamAccnameUtils.buildNameComputationTests(testConfig);
+
```

Comment:
After a shorter example source (e.g. "from title" only), it may be good to show the rendered DOM element(s) that will be generated from that source. E.g. `img` with parent `a href` using 'from title`

PS. can you use `a href` in the parent params, or would it need to be an unlinked anchor? Same question on `img` with an `input type=button` parent.

Reply:
If I’m understanding correctly, this might be pushing my scope even further beyond the [original intent](https://github.com/web-platform-tests/wpt/pull/44965#issue-2172762598)? 😅

It’s [my own damn fault](https://github.com/web-platform-tests/wpt/pull/44965#issuecomment-2063146364), really, but — at least for polishing off this PR — I’ve mainly been focused on generating “simple” subtests for the relatively straightforward conditions specified in [HTML-AAM 4.1](https://www.w3.org/TR/html-aam-1.0/#accessible-name-and-description-computation), particularly in terms of the _absence of potential naming sources_.  Just fleshing out those alone created 500+ subtests. 🤯  And I’m quite certain I’ve even missed some HTML elements as it is.

> PS. can you use a href in the parent params, or would it need to be an unlinked anchor? Same question on img with an input type=button parent.

I’m afraid not, currently — [same answer as above](https://github.com/web-platform-tests/wpt/pull/44965#discussion_r1623038114). But I’d like to add that! I’d like to go even more deeply into interesting subtree-sensitive permutations like `a[href][title] > img`, but maybe that could be a 2.0?

Reply:
Thanks, @cookiecrook. I’ve [marked that file as tentative](https://github.com/web-platform-tests/wpt/pull/44965/commits/ea6ca4cee9da1b02e5547c2ea1c988a135ee7729).

I also found that issue appears to exist already in w3c/html-aam#543, so I left [a comment](https://github.com/w3c/html-aam/issues/543#issuecomment-2415714902) to mention this test.